### ISSUE TITLE
Fix for "manual implementation of Option::map" fails to convert && to &

### ISSUE THREAD

**Issue by cmbartschat** [rust-lang/rust-clippy#13335]
### Summary

When the `manual implementation of Option::map` fix is applied, when there is a `&&Thing` involved, it seems like it doesn't realize it needs a `*` on the return value of `map` so that the returned value is a `&Thing`.

### Reproducer

I tried this code:

```rust
pub struct Thing {
    pub name: String,
}

pub fn filtered_things<'a>(ts: &'a Vec<Thing>) -> Vec<&'a Thing> {
    ts.iter().filter(|t| !t.name.is_empty()).collect()
}

pub fn clippy_repro<'a>(i: usize, things: &'a Vec<Thing>) -> Option<(usize, &'a Thing)> {
    if let Some(v) = filtered_things(&things).get(i) {
        Some((i, v))
    } else {
        None
    }
}

```

I expected to see this happen:

Should fix the code to `filtered_things(things).get(i).map(|v| (i, *v))`

Instead, this happened:

Resulted in `filtered_things(things).get(i).map(|v| (i, v))` (which doesn't compile)

### Version

```text
rustc 1.80.1 (3f5fd8dd4 2024-08-06)
binary: rustc
commit-hash: 3f5fd8dd41153bc5fdca9427e9e05be2c767ba23
commit-date: 2024-08-06
host: aarch64-apple-darwin
release: 1.80.1
LLVM version: 18.1.7
```

**Comment by samueltardieu** [comment]
For reference, the concerned lint is `option_if_let_else`.

**Comment by ada4a** [comment]
The suggestion seems to compile just fine: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=16915be0c3e2c0da396e00d8cd737385

It could be that type elision has become smarter, automatically dereferencing `&&Thing` to `&Thing`.

I'll close this for now, but feel free to re-open if something's not right

**Comment by cmbartschat** [comment]
@ada4a, thanks for taking a look, your playground uses `filtered_things(&things).get(i).map_or(None, |v| Some((i, v)))`, which is not the suggestion I was getting. 

Going back to the original diagnostic, with the original repro case:

https://godbolt.org/z/Y8nGM3he3

```
warning: manual implementation of `Option::map`
  --> <source>:10:5
   |
10 | /     if let Some(v) = filtered_things(&things).get(i) {
11 | |         Some((i, v))
12 | |     } else {
13 | |         None
14 | |     }
   | |_____^ help: try: `filtered_things(&things).get(i).map(|v| (i, v))`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
   = note: `#[warn(clippy::manual_map)]` on by default
```

Looking a bit deeper, I noticed that as of 1.87.0, the suggestion is no longer raised.

Reported (1.86.0): https://godbolt.org/z/Wcad3KTzc
Not reported (1.87.0): https://godbolt.org/z/qsEKPGKo6

In this sample:

```
#![allow(dead_code)]

static  X: i32 = 3;

pub fn clippy_repro1() -> Option<(i32, i32)> {
    let something = Some(&X);
    if let Some(v) = something {
        Some((*v, *v))
    } else {
        None
    }
}

pub fn clippy_repro2() -> Option<(i32, &'static i32)> {
    let something = Some(&X);
    if let Some(v) = something {
        Some((*v, v))
    } else {
        None
    }
}

pub fn clippy_repro3() -> Option<&'static i32> {
    let something = Some(&X);
    if let Some(v) = something {
        Some(v)
    } else {
        None
    }
}

pub fn clippy_repro4() -> Option<i32> {
    let something = Some(&X);
    if let Some(v) = something {
        Some(*v)
    } else {
        None
    }
}
```

https://godbolt.org/z/6Mnq9Mq6j

We don't get warned in cases where we're returning references, which I'm unsure is intended. 

Seems like https://github.com/rust-lang/rust-clippy/pull/14389 would match the timeline of a change made between 14 February, 2025 and 28 March, 2025.

As of right now, the issue I reported does not occur because the suggestion is not raised.

**Comment by cmbartschat** [comment]
Filed https://github.com/rust-lang/rust-clippy/issues/16031

If that issue is fixed, this issue may return.
